java/ql/src/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -680,3 +675,18 @@ private MethodAccess callReturningSameType(Expr ref) {
   ref = result.getQualifier() and
   result.getMethod().getReturnType() = ref.getType()
 }
+
+private import semmle.code.java.dataflow.FlowSources
+
+private RefType entrypointType() {
+  result = any(RemoteFlowSource s | exists(Parameter p | p = s.asParameter())).getType() or
+  result = entrypointType().getASubtype() or
+  result = entrypointType().getAField().getType()
+}
+
+private class EntrypointFieldStep extends TaintTracking::AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
```

Comment:
There's a slight difference between extending `AdditionalTaintStep` and adding to `localAdditionalTaintStep`. The former is meant as a customer-facing extension point and is merely added to all `TaintTracking::Configuration`s, whereas the latter also affects the local taint step relation (which, I'll admit, probably isn't used much). This step is more appropriate to add to `localAdditionalTaintStep`.

The git and PR history doesn't contain any info about whether this difference was intentional (https://github.com/github/codeql/pull/1802), but one reason to keep the difference as-is is that a customer might add a non-local step through `TaintTracking::AdditionalTaintStep`, which definitely shouldn't be included in the local taint step relation.

Reply:
Thank you so much for the context. Knowing these little differences can be a life-saver when facing certain mysteries that don't make any sense otherwise. So I'm adding this to my list of gotchas :-)

I followed your advice in e39445e. I couldn't add it to `localAdditionalTaintExprStep` because `DataFlow::getFieldQualifier` and `Expr` don't work well together, so I added it to `localAdditionalTaintStep`, but please let me know if there's a better place for this.